grpc-server-utils/src/main/java/org/hypertrace/core/grpcutils/server/ThrowableResponseInterceptor.java

```diff
@@ -22,21 +26,49 @@ public void sendMessage(RespT message) {
 
           @Override
           public void close(Status status, Metadata trailers) {
-            if (status.getCode() == Status.Code.UNKNOWN
+            if (status.getCode() == UNKNOWN
                 && status.getDescription() == null
                 && status.getCause() != null) {
               status =
                   Status.INTERNAL
                       .withDescription(status.getCause().getMessage())
                       .withCause(status.getCause());
             }
-            if (!status.isOk() && trailers.keys().isEmpty()) {
-              super.close(status, RequestContext.fromMetadata(headers).buildTrailers());
-            }
-            super.close(status, trailers);
+            super.close(
+                status,
+                collectAndMergeMetadata(RequestContext.fromMetadata(headers), trailers, status));
           }
         };
 
     return next.startCall(wrappedCall, headers);
   }
+
+  private Metadata collectAndMergeMetadata(
+      RequestContext requestContext, Metadata originalTrailers, Status status) {
+    Metadata mergedTrailers = new Metadata();
+    // We build with increasing precedence (since metadata.get() reads last value)
+    mergedTrailers.merge(requestContext.buildTrailers());
+    mergedTrailers.merge(collectAllTrailersFromCause(status));
+    mergedTrailers.merge(originalTrailers);
+
+    return mergedTrailers;
+  }
+
+  private Metadata collectAllTrailersFromCause(Status status) {
+    // Two base cases - either no cause or an unknown cause
+    Throwable cause = status.getCause();
+    if (isNull(cause)) {
+      return new Metadata();
+    }
+    Status statusFromCause = Status.fromThrowable(cause);
+    if (statusFromCause.getCode() == UNKNOWN) {
+      return new Metadata();
+    }
+    // Otherwise, we've found a status so collect any trailers from it and merge them on top of
+    // any trailers we can find from descendents
+    Metadata trailersFromCauseDescendents = this.collectAllTrailersFromCause(statusFromCause);
```

Comment:
Could this collection grow out of bounds affecting memory?

Reply:
Not really a concern. It only recurses for the length of the causal chain, so it's bounded by the size of that. As soon as we run out of causes or hit an unknown cause, we exit recursion (the two base cases above). The causal chain must necessarily be the same depth as this recursion (or longer), and each link in that is a throwable that contains a superset of the data we're copying out.